### DISCUSSION TITLE
main vs master branches

### DISCUSSION THREAD

**Discussion by mices** [community/community#44481] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I've been developing on rails for years but never bothered using github Now that I'm using api's with webhooks I have to develop on heroku and cloud9 so I started using github I've been learning a little but the main vs master confusion is driving me nuts If main replaced master as I've heard it said, why are there main and master branches in my repository When I just run `git push` I get the same code in main but master updates on github Which branch is set to autodeploy?

**Comment by lambdakilo** [comment]
Heroku's default branch is master. Git's default branch is master. Cloud9's default branch is assumed to be master.

GitHub however uses main as their default branch.

If you clone a repository created in GitHub it will most likely use main as it's default branch.

If you push a repository to GitHub or any other code hosting platform which was created with local Git it most likely uses master as it's default branch.

For more info on the branch renaming at GitHub please checkout: https://github.com/github/renaming

**Comment by calebephrem** [comment]


**Reply by metux** [reply]
*rofl* you don't really assume anybody clear-thinking will take this buzzword collection seriously ?
`master` always had been the default on git from day one.

If you prefer wasting your time with weird political activism instead of technical excellence, that's your choice.

**Comment by MuhammadAriyan** [comment]
master was used to be the default branch later it was replaced with main. both branched are of the main author . it's mainly used as default by many repos

**Comment by Sahil-dev7** [comment]
GitHub switched the default branch name from master → main in late 2020 for new repositories.
That’s why you’re seeing both:

Older repos (or tools like Heroku/Cloud9) may still default to master.

Newer repos use main.


In your case, both branches exist, but only one is connected to your deploy pipeline. Heroku (or whatever service you linked) is probably watching master by default. That’s why git push seems to update main locally, while GitHub shows changes under master.

👉 To fix this, you should:

1. Check which branch Heroku is set to auto-deploy from (Dashboard → Deploy tab).


2. Either change Heroku to deploy from main, or update your local repo so you always push to the branch that matches the deploy config.


3. If you don’t need both, delete the extra branch (git branch -d master or git push origin --delete master) so it’s less confusing.

**Comment by sharpjonov** [comment]
Here’s what’s happening:

Originally, Git used master as the default branch name.

In 2020, GitHub changed the default to main for new repositories to make naming more inclusive.

If your local repo was created before that, or cloned from something older, it might still have master as the default branch.

That’s why you might see both main and master — one could be your local branch, and the other is the default on GitHub (or vice versa).

🔧 To fix it:

Run git branch to see which branch you’re currently on.

If you want to standardize on main, run:

git branch -m master main
git push -u origin main


On GitHub, go to your repo’s Settings → Branches → Default branch and make sure it’s set to main.

✅ About auto-deploy:
If you’re using Heroku, it will auto-deploy whatever branch is set as the default branch on GitHub (unless you manually chose a different one in your Heroku app settings).

So, once you make sure main is your default branch both locally and remotely, git push will behave consistently and deploy correctly.

**Reply by Propfend** [reply]
> On GitHub, go to your repo’s Settings → Branches → Default branch and make sure it’s set to main

There is no such configuration anymore.

**Comment by mahirliy55** [comment]
The confusion comes from a naming change in Git and GitHub.

Traditionally, Git used master as the default branch name. However, in mid-2020 GitHub (and many others) switched the default branch name to main for new repositories, to use more inclusive terminology.

Here’s what’s likely happening in your case:

Your local repository was initialized earlier (or cloned from an older repo) where the default branch was still master.

But your GitHub repository (or deployment configuration, like Heroku) is set up with main as the default branch.

So when you run git push, depending on how your local branch is tracked, it might be pushing code to main or master — sometimes both appear if you didn’t align them.

How to fix it

Check which branch you’re currently on:

git branch


If you’re on master and want to rename it to main:

git branch -m master main
git push -u origin main
git push origin --delete master


Confirm on GitHub that main is set as the default branch (Settings → Branches → Default branch).

Update any deployment hooks (like Heroku’s) to deploy from main.

Once both local and remote branches match (main everywhere), the confusion will disappear — and git push will always update the same branch that Heroku is watching for autodeploy.
